src/orthonormal_bases.jl

```diff
@@ -0,0 +1,325 @@
+    M_dim = manifold_dimension(M)
+    N == M_dim || throw(ArgumentError("Incorrect number of basis vectors; expected: $M_dim, given: $N"))
+    for i in 1:N
+        vi_norm = norm(M, x, bvectors[i])
+        isapprox(vi_norm, 1) || throw(ArgumentError("vector number $i is not normalized (norm = $vi_norm)"))
+        for j in i+1:N
+            dot_val = real(inner(M, x, bvectors[i], bvectors[j]))
+            isapprox(dot_val, 0; atol = eps(eltype(x))) || throw(ArgumentError("vectors number $i and $j are not orthonormal (inner product = $dot_val)"))
+        end
+    end
+    return B
+end
+
+function get_coordinates(M::ArrayManifold, x, v, B::AbstractBasis; kwargs...)
+    is_tangent_vector(M, x, v, true; kwargs...)
+    return get_coordinates(M.manifold, x, v, B)
+end
+
+function get_vector(M::ArrayManifold, x, v, B::AbstractBasis; kwargs...)
+    is_manifold_point(M, x, true; kwargs...)
+    size(v) == (manifold_dimension(M),) || error("Incorrect size of vector v")
+    return get_vector(M.manifold, x, v, B)
+end
+
+function _euclidean_basis_vector(x, i)
+    y = zero(x)
+    y[i] = 1
+    return y
+end
+
+function basis(M::Manifold, x, B::ProjectedOrthonormalBasis{:svd, ℝ})
+    S = representation_size(M)
+    PS = prod(S)
+    dim = manifold_dimension(M)
+    # projection
+    # TODO: find a better way to obtain a basis of the ambient space
+    vs = [convert(Vector, reshape(project_tangent(M, x, _euclidean_basis_vector(x, i)), PS)) for i in eachindex(x)]
+    O = reduce(hcat, vs)
+    # orthogonalization
+    # TODO: try using rank-revealing QR here
+    decomp = svd(O)
+    rotated = Diagonal(decomp.S) * decomp.Vt
+    vecs = [collect(reshape(rotated[i,:], S)) for i in 1:dim]
+    # normalization
+    for i in 1:dim
+        i_norm = norm(M, x, vecs[i])
+        vecs[i] /= i_norm
+    end
+    return PrecomputedOrthonormalBasis(vecs)
+end
+
+"""
+    vectors(M::Manifold, x, B::AbstractBasis)
+
+Get the basis vectors of basis `B` of the tangent space at point `x`.
+"""
+function vectors(M::Manifold, x, B::AbstractBasis)
+    error("vectors not implemented for manifold of type $(typeof(M)) a point of type $(typeof(x)) and basis of type $(typeof(B)).")
+end
+
+vectors(::Manifold, x, B::PrecomputedOrthonormalBasis) = B.vectors
+vectors(::Manifold, x, B::PrecomputedDiagonalizingOrthonormalBasis) = B.vectors
+
+# related to DefaultManifold; to be moved to ManifoldsBase.jl in the future
+using ManifoldsBase: DefaultManifold
+function get_coordinates(M::DefaultManifold, x, v, ::ArbitraryOrthonormalBasis)
+    return reshape(v, manifold_dimension(M))
+end
+function get_vector(M::DefaultManifold, x, v, ::ArbitraryOrthonormalBasis)
+    return reshape(v, representation_size(M))
+end
+
+function basis(M::DefaultManifold, x, ::ArbitraryOrthonormalBasis)
+    return PrecomputedOrthonormalBasis([_euclidean_basis_vector(x, i) for i in eachindex(x)])
+end
+
+function basis(M::Manifold, x, B::ProjectedOrthonormalBasis{:gram_schmidt,F}; kwargs...) where F
```

Comment:
This also works for the complex coefficient basis of tangent spaces of the complex Grassman manifold (it's tested).

Reply:
Hmm, it shouldn't, since it's a different inner product, hence a different meaning of "orthonormal". I suspect it's passing because the test only looks at the real part of the inner product. An orthonormal vector with complex coefficients will have a complex inner product between non-identical basis vectors of `0+0im`. For real coefficients, IIRC, the imaginary part is usually non-zero.

Reply:
Right, that's possible. I'll change the signature.